File: examples/human/email/agent.py
"""Budget Envelope Validator — human/email example.

Delivery: stream  (GET /v1/agents/{address}/intents/stream)

Validates a budget request against quarterly allocations before routing
the task to a human approver via email.

Run (Terminal 1 — agent):
    export AXME_API_KEY=<your-api-key>
    export AXME_AGENT_ADDRESS=agent://<org>/<workspace>/budget-envelope-validator
    python agent.py

Run (Terminal 2 — scenario):
    axme scenarios apply examples/human/email/scenario.json --watch

After the agent completes its step, AXME emails the human approver
(the email is set via `axme login` / AXME_USER_EMAIL env var).
The approver clicks the link in the email to approve or reject.
"""
from __future__ import annotations

import logging
import os
import sys
from typing import Any

from axme import AxmeClient, AxmeClientConfig

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s  [%(levelname)s]  %(message)s",
    datefmt="%H:%M:%S",
)
log = logging.getLogger("budget-envelope-validator")

AXME_BASE_URL      = os.environ.get("AXME_BASE_URL", "https://api.cloud.axme.ai")
AXME_API_KEY       = os.environ.get("AXME_API_KEY", "")
AXME_AGENT_ADDRESS = os.environ.get("AXME_AGENT_ADDRESS", "budget-envelope-validator")

if not AXME_API_KEY:
    sys.exit("AXME_API_KEY is required")

# ---------------------------------------------------------------------------
# Budget validation rules
# ---------------------------------------------------------------------------
# Quarterly budget envelopes by category.  Any request within the envelope
# and below the single-approval threshold is approved automatically.
# Requests that exceed the envelope are rejected outright.

_QUARTERLY_ENVELOPES: dict[str, float] = {
    "cloud_infrastructure": 50_000,
    "software_licenses":    20_000,
    "contractor_services":  40_000,
    "hardware":             15_000,
}

_AUTO_APPROVE_THRESHOLD = 5_000   # amounts below this skip human approval


def _validate_budget(payload: dict[str, Any]) -> tuple[bool, str, dict[str, Any]]:
    """Return (within_envelope, reason, metadata)."""
    budget_id = payload.get("budget_id", "?")
    amount    = float(payload.get("amount", 0))
    currency  = payload.get("currency", "USD")
    category  = payload.get("category", "")

    envelope = _QUARTERLY_ENVELOPES.get(category)
    if envelope is None:
        return (
            False,
            f"{budget_id}: unknown category {category!r}; known: {sorted(_QUARTERLY_ENVELOPES)}",
            {},
        )

    if amount > envelope:
        return (
            False,
            f"{budget_id}: {currency} {amount:,.0f} exceeds quarterly envelope of {currency} {envelope:,.0f} for {category!r}",
            {"envelope": envelope, "overage": amount - envelope},
        )

    headroom_pct = round((envelope - amount) / envelope * 100, 1)
    return (
        True,
        f"{budget_id}: {currency} {amount:,.0f} is within {category!r} envelope ({headroom_pct}% headroom)",
        {"envelope": envelope, "headroom_pct": headroom_pct},
    )


def handle_intent(client: AxmeClient, delivery: dict[str, Any]) -> None:
    intent_id = delivery.get("intent_id", "")
    if not intent_id:
        log.warning("received delivery without intent_id, skipping")
        return

    log.info("received intent %s", intent_id)

    try:
        resp   = client.get_intent(intent_id)
        intent = resp.get("intent") or resp
    except Exception as exc:
        log.error("get_intent(%s) failed: %s", intent_id, exc)
        return

    status = (intent.get("lifecycle_status") or intent.get("status") or "").upper()
    if status not in {"CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING"}:
        log.debug("skipping intent %s with status %s", intent_id, status)
        return

    raw_payload       = intent.get("payload") or {}
    effective_payload = raw_payload.get("parent_payload") or raw_payload

    within_envelope, reason, meta = _validate_budget(effective_payload)
    log.info("budget validation for %s: ok=%s  reason=%s", intent_id, within_envelope, reason)

    try:
        if within_envelope:
            client.resume_intent(
                intent_id,
                {
                    "action":           "complete",
                    "within_envelope":  True,
                    "reason":           reason,
                    "envelope":         meta.get("envelope"),
                    "headroom_pct":     meta.get("headroom_pct"),
                },
                owner_agent=AXME_AGENT_ADDRESS,
            )
        else:
            client.resume_intent(
                intent_id,
                {
                    "action":          "fail",
                    "within_envelope": False,
                    "reason":          reason,
                },
                owner_agent=AXME_AGENT_ADDRESS,
            )
        log.info("resumed intent %s (within_envelope=%s)", intent_id, within_envelope)
    except Exception as exc:
        log.error("resume_intent(%s) failed: %s", intent_id, exc)


def main() -> None:
    client = AxmeClient(AxmeClientConfig(api_key=AXME_API_KEY, base_url=AXME_BASE_URL))
    log.info(
        "budget-envelope-validator starting  address=%s  binding=stream",
        AXME_AGENT_ADDRESS,
    )

    try:
        for delivery in client.listen(AXME_AGENT_ADDRESS, timeout_seconds=None):
            handle_intent(client, delivery)
    except KeyboardInterrupt:
        log.info("shutting down")


if __name__ == "__main__":
    main()

File: examples/human/email/typescript/agent.ts
/**
 * Budget Envelope Validator — human/email example (TypeScript).
 *
 * Validates budget requests against quarterly allocations before
 * routing to a human approver via email.
 *
 * Run:
 *   export AXME_API_KEY=<your-api-key>
 *   export AXME_AGENT_ADDRESS=budget-envelope-validator
 *   npx tsx examples/human/email/agent.ts
 */
import { AxmeClient } from "@axme/axme";

const AXME_BASE_URL = process.env.AXME_BASE_URL ?? "https://api.cloud.axme.ai";
const AXME_API_KEY = process.env.AXME_API_KEY ?? "";
const AXME_AGENT_ADDRESS = process.env.AXME_AGENT_ADDRESS ?? "budget-envelope-validator";

if (!AXME_API_KEY) { console.error("AXME_API_KEY is required"); process.exit(1); }

// --- Business logic ---
const QUARTERLY_ENVELOPES: Record<string, number> = {
  cloud_infrastructure: 50_000, software_licenses: 20_000,
  contractor_services: 40_000, hardware: 15_000,
};

function validateBudget(payload: Record<string, unknown>): { withinEnvelope: boolean; reason: string; meta: Record<string, unknown> } {
  const budgetId = String(payload.budget_id ?? "?");
  const amount = Number(payload.amount ?? 0);
  const currency = String(payload.currency ?? "USD");
  const category = String(payload.category ?? "");

  const envelope = QUARTERLY_ENVELOPES[category];
  if (envelope === undefined) {
    return { withinEnvelope: false, reason: `${budgetId}: unknown category '${category}'; known: ${Object.keys(QUARTERLY_ENVELOPES).sort()}`, meta: {} };
  }
  if (amount > envelope) {
    return { withinEnvelope: false, reason: `${budgetId}: ${currency} ${amount.toLocaleString()} exceeds quarterly envelope of ${currency} ${envelope.toLocaleString()} for '${category}'`, meta: { envelope, overage: amount - envelope } };
  }

  const headroomPct = Math.round((envelope - amount) / envelope * 1000) / 10;
  return { withinEnvelope: true, reason: `${budgetId}: ${currency} ${amount.toLocaleString()} is within '${category}' envelope (${headroomPct}% headroom)`, meta: { envelope, headroom_pct: headroomPct } };
}

// --- Agent loop ---
const client = new AxmeClient({ apiKey: AXME_API_KEY, baseUrl: AXME_BASE_URL });
console.log(`budget-envelope-validator starting  address=${AXME_AGENT_ADDRESS}  binding=stream`);

try {
  for await (const delivery of client.listen(AXME_AGENT_ADDRESS)) {
    const intentId = String(delivery.intent_id ?? "");
    if (!intentId) continue;
    console.log(`received intent ${intentId}`);

    let intent: Record<string, unknown>;
    try {
      const resp = await client.getIntent(intentId);
      intent = (resp.intent ?? resp) as Record<string, unknown>;
    } catch (e) { console.error(`get_intent(${intentId}) failed:`, e); continue; }

    const status = String(intent.lifecycle_status ?? intent.status ?? "").toUpperCase();
    if (!["CREATED", "DELIVERED", "ACKNOWLEDGED", "IN_PROGRESS", "WAITING"].includes(status)) continue;

    const rawPayload = (intent.payload ?? {}) as Record<string, unknown>;
    const effectivePayload = (rawPayload.parent_payload ?? rawPayload) as Record<string, unknown>;
    const { withinEnvelope, reason, meta } = validateBudget(effectivePayload);

    console.log(`budget validation for ${intentId}: ok=${withinEnvelope}  reason=${reason}`);

    try {
      if (withinEnvelope) {
        await client.resumeIntent(intentId, {
          action: "complete", within_envelope: true, reason,
          envelope: meta.envelope, headroom_pct: meta.headroom_pct,
        }, { ownerAgent: AXME_AGENT_ADDRESS });
      } else {
        await client.resumeIntent(intentId, {
          action: "fail", within_envelope: false, reason,
        }, { ownerAgent: AXME_AGENT_ADDRESS });
      }
      console.log(`resumed intent ${intentId} (within_envelope=${withinEnvelope})`);
    } catch (e) { console.error(`resume_intent(${intentId}) failed:`, e); }
  }
} catch (e) {
  if ((e as Error).name !== "AbortError") throw e;
}
